File: server/src/main/kotlin/com/xpromus/onebike_backend/rider/mapper/RiderMapper.kt
package com.xpromus.onebike_backend.rider.mapper

import com.xpromus.onebike_backend.nation.Nation
import com.xpromus.onebike_backend.nation.dto.NationDescriptorDto
import com.xpromus.onebike_backend.placement.dto.PlacementDescriptorDto
import com.xpromus.onebike_backend.rider.Rider
import com.xpromus.onebike_backend.rider.dto.GetRiderDto
import com.xpromus.onebike_backend.rider.dto.GetRiderWithChildrenDto
import com.xpromus.onebike_backend.rider.dto.PostRiderDto
import com.xpromus.onebike_backend.rider.dto.PutRiderDto
import com.xpromus.onebike_backend.rider.dto.RiderDescriptorDto
import com.xpromus.onebike_backend.team.Team
import com.xpromus.onebike_backend.team.dto.TeamDescriptorDto

fun Rider.toGetRiderDto(
    placementIds: List<Long>,
    nationId: Long,
    teamId: Long?
): GetRiderDto {
    return GetRiderDto(
        id = id!!,
        firstName = firstName,
        lastName = lastName,
        dateOfBirth = dateOfBirth,
        placementIds = placementIds,
        nationId = nationId,
        teamId = teamId
    )
}

fun Rider.toGetRiderWithChildrenDto(
    placements: List<PlacementDescriptorDto>,
    nation: NationDescriptorDto,
    team: TeamDescriptorDto?
): GetRiderWithChildrenDto {
    return GetRiderWithChildrenDto(
        id = id!!,
        firstName = firstName,
        lastName = lastName,
        dateOfBirth = dateOfBirth,
        placements = placements,
        nation = nation,
        team = team,
    )
}

fun Rider.toRiderDescriptorDto(): RiderDescriptorDto {
    return RiderDescriptorDto(
        id = id!!,
        firstName = firstName,
        lastName = lastName,
        dateOfBirth = dateOfBirth
    )
}

fun PutRiderDto.toEntity(
    original: Rider,
    nation: Nation,
    team: Team?,
): Rider {
    return Rider(
        id = original.id,
        firstName = firstName,
        lastName = lastName,
        dateOfBirth = dateOfBirth,
        placements = original.placements,
        nation = nation,
        team = team,
    )
}

fun PutRiderDto.toNewEntity(
    nation: Nation,
    team: Team?,
): Rider {
    return Rider(
        firstName = firstName,
        lastName = lastName,
        dateOfBirth = dateOfBirth,
        nation = nation,
        team = team,
    )
}

fun PostRiderDto.toNewEntity(
    nation: Nation,
    team: Team?,
): Rider {
    return Rider(
        firstName = firstName,
        lastName = lastName,
        dateOfBirth = dateOfBirth,
        nation = nation,
        team = team,
    )
}

File: admin_dashboard/src/lib/middleware/rider.mapper.ts
import type { RiderWithChildren, RiderWithIds } from "$lib/types/client/rider.types";
import type { GetRiderDto, GetRiderWithChildrenDto } from "$lib/types/server/rider.types";

export const toRiderWithIds = (
    rider: GetRiderDto
): RiderWithIds => {
    return {
        id: rider.id,
        firstName: rider.firstName,
        lastName: rider.lastName,
        dateOfBirth: new Date(rider.dateOfBirth),
        placementIds: rider.placementIds,
        nationId: rider.nationId
    } satisfies RiderWithIds;
}

export const toRiderWithChildren = (
    rider: GetRiderWithChildrenDto
): RiderWithChildren => {
    return {
        id: rider.id,
        firstName: rider.firstName,
        lastName: rider.lastName,
        dateOfBirth: new Date(rider.dateOfBirth),
        placements: rider.placements,
        nation: rider.nation,
        team: rider.team
    } satisfies RiderWithChildren;
}

export const riderWithChildrenToRiderWithIds = (
    rider: RiderWithChildren
): RiderWithIds => {
    return {
        id: rider.id,
        firstName: rider.firstName,
        lastName: rider.lastName,
        dateOfBirth: rider.dateOfBirth,
        placementIds: rider.placements.map(
            (placement) => {
                return placement.id
            }
        ),
        nationId: rider.nation.id
    } satisfies RiderWithIds
}
